packages/controllers/src/multichain/MultiChainController.ts

```diff
@@ -0,0 +1,453 @@
+ * @returns The parsed chain ID.
+ */
+export function parseChainId(chainId: ChainId): {
+  namespace: NamespaceId;
+  reference: string;
+} {
+  if (!isChainId(chainId)) {
+    throw new Error('Invalid chain ID.');
+  }
+
+  const match = CHAIN_ID_REGEX.exec(chainId);
+  if (!match?.groups) {
+    throw new Error('Invalid chain ID.');
+  }
+
+  return {
+    namespace: match.groups.namespace,
+    reference: match.groups.reference,
+  };
+}
+
+/**
+ * Parse an account ID to an object containing the chain, chain ID and address.
+ * This validates the account ID before parsing it.
+ *
+ * @param accountId - The account ID to validate and parse.
+ * @returns The parsed account ID.
+ */
+export function parseAccountId(accountId: AccountId): {
+  chain: { namespace: NamespaceId; reference: string };
+  chainId: ChainId;
+  address: string;
+} {
+  if (!isAccountId(accountId)) {
+    throw new Error('Invalid account ID.');
+  }
+
+  const match = ACCOUNT_ID_REGEX.exec(accountId);
+  if (!match?.groups) {
+    throw new Error('Invalid account ID.');
+  }
+
+  return {
+    address: match.groups.accountAddress,
+    chainId: match.groups.chainId as ChainId,
+    chain: {
+      namespace: match.groups.namespace,
+      reference: match.groups.reference,
+    },
+  };
+}
+
+// TODO(ritave): Support for legacy ethereum operations, not just snaps
+export class MultiChainController extends BaseController<
+  typeof controllerName,
+  MultiChainControllerState,
+  MultiChainControllerMessenger
+> {
+  private notify: Notify;
+
+  constructor({ messenger, notify }: MultiChainControllerArgs) {
+    super({
+      messenger,
+      metadata: {
+        sessions: { persist: false, anonymous: false },
+      },
+      name: controllerName,
+      state: defaultState,
+    });
+
+    this.notify = notify;
+  }
+
+  getSession(origin: string): SessionData | undefined {
+    return this.state.sessions[origin];
+  }
+
+  async closeSession(origin: string): Promise<void> {
+    const session = this.getSession(origin);
+    assert(session, new Error('No session to close'));
+
+    await this.notify(origin, { method: 'multichainHack_metamask_disconnect' });
+
+    this.update((state) => {
+      delete state.sessions[origin];
+    });
+
+    await Promise.all(
+      Object.values(session.handlingSnaps).map((snapId) =>
+        this.messagingSystem.call('SnapController:onSessionClose', snapId),
+      ),
+    );
+  }
+
+  async onConnect(
+    origin: string,
+    connection: ConnectArguments,
+  ): Promise<Session> {
+    const existingSession = this.getSession(origin);
+    if (existingSession) {
+      await this.closeSession(origin);
+    }
+
+    const snaps = await this.messagingSystem.call('SnapController:getAll');
+    const filteredSnaps = snaps.filter((snap) => snap.enabled && !snap.blocked);
+
+    const availableNamespaces = await filteredSnaps.reduce<
+      Promise<Record<SnapId, Record<NamespaceId, Namespace>>>
+    >(async (previousPromise, snap) => {
```

Comment:
Why is this a Promise? It's coming from the PermissionController or defaultPermissions.

Also it's linear. The event queue can manage the load for us.

```typescript
const availableNamespaces = Object.fromEntries(await Promise.all(filteredSnaps.map(async (snap) => [snap.id, await this.snapToNamespaces(snap)]));
```

Reply:
> Why is this a Promise?
Future-proofing in case messaging isn't synchronous in the very near future.

Reply:
But good point on turning it into a `Promise.all`